platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencySubstitutionRulesIntegrationTest.groovy

```diff
@@ -1612,4 +1658,64 @@ Required by:
             }
         }
     }
+
+    @Issue("https://github.com/gradle/gradle/issues/7594")
+    def "can substitute module with project and use withoutArtifactSelectors"() {
+        mavenRepo.module("com.external", "libB", "1.0")
+            .dependsOn("com.external", "libC", "1.0", "type")
+            .publish()
+
+        settingsFile << """
+            include("libC")
+        """
+
+        buildFile << """
+            plugins {
+                id("java-library")
+            }
+
+            ${mavenTestRepository()}
+
+            dependencies {
+                implementation("com.external:libB:1.0")
+            }
+
+            configurations.runtimeClasspath.resolutionStrategy.dependencySubstitution {
+                def sub = substitute(module("com.external:libC:1.0"))
+                    .using(project(":libC"))
+
+                if (${withoutArtifactSelectors}) {
+                    sub.withoutArtifactSelectors()
+                }
+            }
+
+            tasks.register("resolve") {
+                def files = configurations.runtimeClasspath.incoming.files
+                doLast {
+                    println(files*.name)
+                }
+            }
+        """
+
+        file("libC/build.gradle") << """
+            plugins {
+                id("java-library")
+            }
+        """
+
+        expect:
+        if (withoutArtifactSelectors) {
+            succeeds(":resolve")
+            outputContains("[libB-1.0.jar, libC.jar]")
+        } else {
+            fails(":resolve")
+            failure.assertHasCause("Could not find libC.type (project :libC)")
```

Comment:
Is there any way to make this work without `withoutArtifactSelectors`?  Can libC register an artifact that we'd find with the right type?

Reply:
For this to work, the variant that we select from `libC` must have an artifact with type `type`. 

So we could theoretically add a new artifact to libC's runtimeElements and this should pass. 

Though, that was not the intent of this code here. I wanted to show that adding `withoutArtifactSelectors` does indeed "fix" the situation here